### ISSUE TITLE
Unable to use InstrumentedAttribute to value mappings in mysql/mariadb on_duplicate_key_update

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#12117]
### Discussed in https://github.com/sqlalchemy/sqlalchemy/discussions/12116

<div type='discussions-op-text'>

<sup>Originally posted by **MajorTanya** November 21, 2024</sup>
Hi everyone, I am encountering a (to me) strange issue/inconsistency in how I expect SQLAlchemy to work with the `on_duplicate_key_update` function for MySQL/MariaDB dialects.

(I'm only a beginner at using SQLAlchemy, so this may be a misunderstanding on my end.)

### Versions used
- MariaDB Server 11.5.2

PyPI packages:
- `mariadb==1.1.11`
- `SQLAlchemy==2.0.36`

I discovered that I can use a mapping of the model class to the new value directly, which I appreciate tremendously. I'm always wary of typos and outdated kwargs not matching column names, etc. It appears to work tremendously from my trying it out and I wanted to use it everywhere just out of consistency. But in trying to do so in an `on_duplicate_key_update` clause, I was faced with an exception.

### Reproduction
An almost minimal reproduction can look like this:

```python
from sqlalchemy import create_engine, Integer, insert, delete
from sqlalchemy.dialects.mysql import insert as mariadb_insert
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    Session,
    mapped_column,
)


class Base(DeclarativeBase):
    pass


class MyThing(Base):
    __tablename__ = "thing_table"

    number: Mapped[int] = mapped_column(Integer, primary_key=True)


db_host = "localhost"
db_user = "my_user"
db_pass = "my_password"

engine = create_engine(
    f"mariadb+mariadbconnector://{db_user}:{db_pass}@{db_host}/my_db",
    echo=True,
)

Base.metadata.create_all(engine)

with Session(engine) as session:
    session.execute(delete(MyThing))  # just clearing out for repeated use
    session.execute(insert(MyThing).values({MyThing.number: 1}))
    session.commit()

    colliding_upsert = (
        mariadb_insert(MyThing)
        .values({MyThing.number: 1})
        .on_duplicate_key_update({MyThing.number: 2})
    )

    compiled = colliding_upsert.compile(compile_kwargs={"literal_binds": True})
    print("Compiled statement:", compiled, "||")

    session.execute(colliding_upsert)

    session.commit()
```

I expected this to apply the change as expressed in the `on_duplicate_key_update` clause: I wanted it to set the `MyThing.number` column to 1.

Instead, I am faced with the following exception (I can provide the full traceback if needed but this appears to encapsulate both the MariaDB exception and the SQLAlchemy exception):

```
sqlalchemy.exc.ProgrammingError: (mariadb.ProgrammingError) You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
[SQL: INSERT INTO thing_table (number) VALUES (?) ON DUPLICATE KEY UPDATE ]
[parameters: (1,)]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

It appears to me that the `ON DUPLICATE KEY UPDATE` clause was not generated fully/correctly by SQLAlchemy. If I change the InstrumentedAttribute to value mapping input to the kwarg approach like this

```diff
    colliding_upsert = (
        mariadb_insert(MyThing)
        .values({MyThing.number: 1})
-       .on_duplicate_key_update({MyThing.number: 2})
+       .on_duplicate_key_update(number=2)
    )
```

the entire script finishes without exceptions and the table is updated as expected.

### Compiled statements

For the failing approach, the compiled statement looks like this:

```txt
Compiled statement: INSERT INTO thing_table (number) VALUES (1) ON DUPLICATE KEY UPDATE  ||
```
(the double pipe is from my printing in the example to show the end of the line, so this is the whole entire statement as generated by SQLAlchemy)

For the succeeding kwarg approach, the same line outputs the following statement:
```txt
Compiled statement: INSERT INTO thing_table (number) VALUES (1) ON DUPLICATE KEY UPDATE number = %s ||
```

This appears to show that the mapping approach, while obviously supported for "normal" inserts (see the first insertion being made, for example), is not supported by the on_duplicate_key_update clause of the MySQL/MariaDB dialect insert function.

### The Question

All of this boils down to me wondering whether this non-support is intentional (maybe there are limitations internally/with the DBAPI/etc?) or maybe just an oversight/non-priority.

I didn't open this as an issue because I presume that I'm missing some elephant in some room here. Maybe the way I use the InstrumentedAttribute to value mappings is not even intended and that's why it is not supported in this case. I would appreciate if someone could enlighten me on this situation.

Finally, I hope this wasn't too wordy, I just wanted to make sure I could post as full an example as possible while also keeping it as short as possible and not losing any details. Thank you for your time.</div>

**Comment by sqla-tester** [comment]
**Mingyu Park** has proposed a fix for this issue in the **main** branch:

**Fixes: #12117 Unable to use InstrumentedAttribute to value mappings in mysql/mariadb on_duplicate_key_update** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5674

**Comment by sqla-tester** [comment]
**Mingyu Park** has proposed a fix for this issue in the **main** branch:

**Unable to use InstrumentedAttribute to value mappings in mysql/mariadb on_duplicate_key_update** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5674

**Comment by sqla-tester** [comment]
**Mingyu Park** has proposed a fix for this issue in the **rel_2_0** branch:

**Unable to use InstrumentedAttribute to value mappings in mysql/mariadb on_duplicate_key_update** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5676
